### PR TITLE
Optimize our String/&str/Vec<u8>/&[u8] IntoDatum conversions

### LINKED ISSUE #941: Investigate if String::into_datum() and &str::into_datum() can be optimised
Calling String::into_datum() and &str::into_datum() seem to take a lot of time. 

Running a trace in a profiler when selecting 300 million rows, String::into_datum() showed up as taking 25-30% of the total query time. For a "simple" conversion to a Datum this seemed quite excessive.

This issue is to investigate if this conversion can be optimised.

### PR DESCRIPTION
**Pull request by eeeebbbbrrrr**
We now directly palloc the `pg_sys::varlena` buffer, set its size, and essentially just memcpy the Rust bytes to the Postgres-allocated memory.

closes #941

### CODE REVIEW

**Comment by eeeebbbbrrrr** [comment]
See my comment over on #941 about timing numbers:  https://github.com/tcdi/pgx/issues/941#issuecomment-1345613328

File: pgx/src/datum/into.rs [hunk 1]
```diff
@@ -333,11 +325,31 @@ impl<'a> IntoDatum for &'a crate::cstr_core::CStr {
...
     #[inline]
     fn into_datum(self) -> Option<pg_sys::Datum> {
-        let varlena = rust_byte_slice_to_bytea(&self);
-        if varlena.is_null() {
-            None
-        } else {
-            Some(varlena.into_pg().into())
+        let len = pg_sys::VARHDRSZ + self.len();
+        unsafe {
+            // SAFETY:  palloc gives us a valid pointer if if there's not enough memory it'll raise an error
```
Comment (BradyBonnette) [inline pgx/src/datum/into.rs:330]: `if if`

File: pgx/src/datum/into.rs [hunk 2]
```diff
@@ -333,11 +325,31 @@ impl<'a> IntoDatum for &'a crate::cstr_core::CStr {
...
+                .cast::<pg_sys::varattrib_4b>()
+                .as_mut()
+                .unwrap_unchecked()
+                .va_4byte
+                .as_mut();
+            varattrib_4b.va_header = <usize as TryInto<u32>>::try_into(len)
+                .expect("Rust string too large for a Postgres varlena datum")
+                << 2u32;
+
+            // SAFETY: src and dest pointers are valid and are exactly `self.len()` bytes long
```
Comment (workingjubilee) [inline pgx/src/datum/into.rs:345]: hidden implicit assumption:
```suggestion
            // SAFETY: src and dest pointers are valid, exactly `self.len()` bytes long,
            // and the `dest` was freshly allocated, thus non-overlapping
```

File: pgx/src/datum/into.rs [hunk 3]
```diff
@@ -333,11 +325,31 @@ impl<'a> IntoDatum for &'a crate::cstr_core::CStr {
...
+            // allocated thanks to our call to `palloc` above
+            let varattrib_4b = varlena
+                .cast::<pg_sys::varattrib_4b>()
+                .as_mut()
+                .unwrap_unchecked()
+                .va_4byte
+                .as_mut();
+            varattrib_4b.va_header = <usize as TryInto<u32>>::try_into(len)
+                .expect("Rust string too large for a Postgres varlena datum")
+                << 2u32;
```
Comment (workingjubilee) [inline pgx/src/datum/into.rs:343]: Intended resulting bitmask is `0bxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx00`, correct? And the two LSB will be used by Postgres for the varlena tag?

**Review by workingjubilee — APPROVED** [review]

File: pgx/src/datum/into.rs (hunk 3 above)
Comment (eeeebbbbrrrr) [reply pgx/src/datum/into.rs:343]: This is the same as Postgres' `#define SET_VARSIZE_4B` (which we do have in `pgx/src/varlena.rs`), but I wanted that `.expect()` on the size.  I'll at least comment this

**Pull request merged**